CMake: enable tests on cross-compilation jobs

janvrany opened the pull request:
This PR provides CMake-based support for running tests under simlator such as QEMU.
This can be used to run tests as part of cross-compilation pipelines - this PR enables it on RISC-V. 

This is an alternative to what has been proposed an implemented by @fjeremic in PR #5753 

As of now, not all tests pass (for example, some port tests fail but I suppose that's because of the use of emulator). 

I did not spend time fixing them yet - the aim of this (draft) PR is to demonstrate an alternative and spark
discussion. 

FYI: @fjeremic , @AdamBrousseau

janvrany commented:
@genie-omr build riscv

fjeremic commented:
@genie-omr build riscv

fjeremic commented on buildenv/jenkins/omrbuild.groovy:
```diff
@@ -174,11 +174,11 @@ def SPECS = [
             ]
         ],
-        'test' : false
+        'test' : true
```
Could we add:

```
        'testArgs' : '',
        'junitPublish' : true
```

to be consistent with the other definitions?

fjeremic commented on cmake/modules/OmrTest.cmake:
```diff
@@ -0,0 +1,49 @@
+	endif()
+endif()
+
+function(omr_add_test)
+	cmake_parse_arguments(
+        opt    # prefix of output variables
```
We have some mix of tabs and spaces here. I forget what we use in other CMake files but let's do the same as them.

fjeremic commented on cmake/modules/OmrTest.cmake:
```diff
@@ -0,0 +1,49 @@
+endif()
+
+function(omr_add_test)
+	cmake_parse_arguments(
+        opt    # prefix of output variables
+        ""     # list of names of the boolean arguments (only defined ones will be true)
+        "NAME;COMMAND" # list of names of mono-valued arguments
+        "" # list of names of multi-valued arguments (output variables are lists)
+        ${ARGN}# arguments of the function to parse, here we take the all original ones
+    )
```
The comments repeat what the [documentation already describes](https://cmake.org/cmake/help/latest/command/cmake_parse_arguments.html). I don't think they're necessary.

```suggestion
	cmake_parse_arguments(opt "" "NAME;COMMAND" "" ${ARGN})
```

fjeremic commented on example/CMakeLists.txt:
```diff
@@ -50,7 +50,7 @@ if(OMR_GC)
 	)
 
-	add_test(NAME gcexample COMMAND gcexample)
+	add_test(NAME gcexample COMMAND $<TARGET_FILE:gcexample>)
```
Why does this not go through `omr_add_test` like the others?

fjeremic commented:
@genie-omr build riscv

fjeremic commented:
```
10:04:08  -- Found LibElf: /usr/lib/x86_64-linux-gnu/libelf.so  
10:04:08  -- Could NOT find LibZ (missing: LIBZ_LIBRARY) 
10:04:08  CMake Error at /usr/local/share/cmake-3.11/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
10:04:08    Could NOT find LibDwarf (missing: LIBDWARF_LIBRARY)
10:04:08  Call Stack (most recent call first):
10:04:08    /usr/local/share/cmake-3.11/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
10:04:08    cmake/modules/FindLibDwarf.cmake:107 (find_package_handle_standard_args)
10:04:08    ddr/lib/ddr-scanner/CMakeLists.txt:47 (find_package)
```

```
jenkins@deb10-x64-1:~$ sudo apt-get install libdwarf-dev
Reading package lists... Done
Building dependency tree
Reading state information... Done
libdwarf-dev is already the newest version (20180809-1).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.

jenkins@deb10-x64-1:~$ ls -alrt /usr/lib/x86_64-linux-gnu/ | grep dw
-rw-r--r--  1 root root   419936 Aug 29  2018 libdwarf.so.1.0.0
lrwxrwxrwx  1 root root       17 Aug 29  2018 libdwarf.so.1 -> libdwarf.so.1.0.0
lrwxrwxrwx  1 root root       17 Aug 29  2018 libdwarf.so -> libdwarf.so.1.0.0
-rw-r--r--  1 root root   638954 Aug 29  2018 libdwarf.a
lrwxrwxrwx  1 root root       14 May 28  2019 libdw.so.1 -> libdw-0.176.so
lrwxrwxrwx  1 root root       10 May 28  2019 libdw.so -> libdw.so.1
-rw-r--r--  1 root root   721718 May 28  2019 libdw.a
-rw-r--r--  1 root root   347776 May 28  2019 libdw-0.176.so
```

Not sure why DDR module can't find libdwarf. I've installed the libdwarf-dev and libdw-dev packages. @keithc-ca this is a Debian system. Any ideas what it may be looking for or what we are missing?

fjeremic commented:
@genie-omr build riscv

keithc-ca commented:
Building DDR only works for non-cross-compile configurations. It may be looking for libdwarf for the target, when what is needed is the host library.

To enable using it in cross-compiles, it would either have to avoid using the port library, etc. or we'd have to build those dependencies for the host as well. I think the former is a better approach, which would see DDR reclassified as a 'tool' (next to hookgen, tracegen, etc.).

fjeremic commented:
Thanks for the quick reply. I don't think this is worth our time. We'll get DDR testing on all other native platforms anyway. I think we can safely just disable DDR for the RISC-V cross compile pipeline. @janvrany could you add `-DOMR_DDR=OFF` to your CMake arguments list?

janvrany commented:
@keithc-ca , @fjeremic, if DDR is not supported when cross-compiling, why don't we automagically set `OMR_DDR=OFF` when cross-compiling?

keithc-ca commented:
> ... why don't we automagically set `OMR_DDR=OFF` when cross-compiling?

That seems like a reasonable plan. I don't know if we can tell with cmake whether the user explicitly enabled DDR, but if so, we should not be ignoring that request; perhaps failing in the cross-compile case.

janvrany commented:
@keithc-ca: Sure. I think I know how to do that, will have a look tomorrow.

janvrany commented:
@keithc-ca, @fjeremic: see my attempt to address DDR issue in dfea0b1
@fjeremic: addressed your comments in d0c56ea. I'll squash this commit into 593a9bc once (if ever :-) this is ready to be merged.

janvrany commented:
@genie-omr build riscv

fjeremic commented on cmake/modules/OmrDDRSupport.cmake:
```diff
@@ -31,6 +31,22 @@ include(ExternalProject)
 # Make sure we have performed platform detection
 include(OmrPlatform)
 
+# DDR is not supported when cross-compiling. Here we automagically
+# enable DDR for native builds and disable for cross-compilation.
+#
+# If OMR_DDR is explicitly set, respect that request (if possible).
+if(NOT DEFINED OMR_DDR)
+	if(CMAKE_CROSSCOMPILING)
+		set(OMR_DDR OFF)
+	else()
+		set(OMR_DDR ON)
+	endif()
+endif()
+
+if(OMR_DDR AND CMAKE_CROSSCOMPILING)
+	message(FATAL_ERROR "DDR is not supported when cross-compiling.")
+endif()
+
```
@keithc-ca @dnakamura tagging for a review here. Is this implementation ok with you?

fjeremic commented on buildenv/jenkins/omrbuild.groovy:
```diff
@@ -174,11 +174,13 @@ def SPECS = [
             ]
         ],
-        'test' : false
+        'test' : true
```
I think we're missing `,` here which is why the build fails.

keithc-ca commented on buildenv/jenkins/omrbuild.groovy:
```diff
@@ -179,6 +179,8 @@ def SPECS = [
         ],
         'test' : true
+        'testArgs' : '',
+        'junitPublish' : true
```
Missing comma.

keithc-ca commented on cmake/caches/Travis.cmake:
```diff
@@ -20,7 +20,6 @@
 #############################################################################
 
 set(OMR_COMPILER ON CACHE BOOL "")
-set(OMR_DDR ON CACHE BOOL "")
```
We don't want to disable building DDR for other platforms.

keithc-ca commented on cmake/caches/Windows.cmake:
```diff
@@ -20,7 +20,6 @@
 ###############################################################################
 
 set(OMR_COMPILER ON CACHE BOOL "")
-set(OMR_DDR ON CACHE BOOL "")
```
We don't want to disable building DDR for other platforms.

keithc-ca commented on cmake/config.cmake:
```diff
@@ -33,7 +33,6 @@ set(OMR_EXAMPLE ON CACHE BOOL "Enable the Example application")
 ###
 
 set(OMR_TOOLS ON CACHE BOOL "Enable the native build tools")
-set(OMR_DDR OFF CACHE BOOL "Enable DDR")
```
We don't want to disable building DDR by default.

keithc-ca commented on fvtest/compilertest/CMakeLists.txt:
```diff
@@ -124,5 +124,5 @@ target_link_libraries(compilertest
 
 if (NOT OMR_HOST_OS STREQUAL "aix")
-	add_test(NAME CompilerTest COMMAND compilertest --gtest_output=xml:${CMAKE_CURRENT_BINARY_DIR}/compilertest-results.xml)
+	add_test(NAME CompilerTest COMMAND $<TARGET_FILE:compilertest> --gtest_output=xml:${CMAKE_CURRENT_BINARY_DIR}/compilertest-results.xml)
```
i think this should use `omr_add_test`.

keithc-ca commented on fvtest/gc_api_test/CMakeLists.txt:
```diff
@@ -34,7 +34,7 @@ function(omr_add_gc_test testname)
-	add_test(
+	omr_add_test(
 		NAME    "${testname}"
 		COMMAND "${testname}" "--gtest_output=xml:${CMAKE_CURRENT_BINARY_DIR}/${testname}-results.xml"
```
Perhaps this should say `COMMAND "$<TARGET_FILE:${testname}>"`?

keithc-ca commented on fvtest/jitbuildertest/CMakeLists.txt:
```diff
@@ -60,5 +60,5 @@ set_property(TARGET jitbuildertest PROPERTY FOLDER fvtest)
 # until #4764 is addressed. Once done, tests should be properly skipped.
 if(NOT OMR_HOST_ARCH STREQUAL "riscv")
-	add_test(NAME JitBuilderTest COMMAND jitbuildertest --gtest_output=xml:${CMAKE_CURRENT_BINARY_DIR}/jitbuildertest-results.xml)
+	add_test(NAME JitBuilderTest COMMAND $<TARGET_FILE:jitbuildertest> --gtest_output=xml:${CMAKE_CURRENT_BINARY_DIR}/jitbuildertest-results.xml)
```
Shouldn't this use `omr_add_test`?

keithc-ca commented on fvtest/threadtest/CMakeLists.txt:
```diff
@@ -79,8 +79,8 @@ endif()
+omr_add_test(NAME threadSetAttrThreadWeightTest COMMAND $<TARGET_FILE:omrthreadtest> --gtest_output=xml:${CMAKE_CURRENT_BINARY_DIR}/omrthreadtest-results.xml --gtest_also_run_disabled_tests --gtest_filter=ThreadCreateTest.DISABLED_SetAttrThreadWeight)
 if(OMR_HOST_OS STREQUAL "linux")
-	add_test(NAME threadRealtimeTest COMMAND omrthreadtest --gtest_output=xml:${CMAKE_CURRENT_BINARY_DIR}/omrthreadtest-results.xml --gtest_filter=ThreadCreateTest.* -realtime)
+	add_test(NAME threadRealtimeTest COMMAND $<TARGET_FILE:omrthreadtest> --gtest_output=xml:${CMAKE_CURRENT_BINARY_DIR}/omrthreadtest-results.xml --gtest_filter=ThreadCreateTest.* -realtime)
```
I think this should use `omr_add_test`.

janvrany commented:
@genie-omr build riscv

janvrany commented:
@keithc-ca:
> We don't want to disable building DDR for other platforms.

I'm sorry I'm confused. On what platforms do we want to have DDR by default? 

The thing is, we're using the same cache file for both native and cross-compilation builds, so if we set it ON somewhere, the check will fail. 

What can be done is leave cache files and `config.cmake` intact and when cross-compiling force it back to OFF. And maybe issue a warning? But then, we would get this warning (which is everyone learns to ignore) each time we cross-compile unless we add `-DOMR_DDR=OFF` at CI job's `cmake` command line (and the point of this exercise was to avoid that, making it easier / simpler to use).

Or am I missing something?

keithc-ca commented:
> > We don't want to disable building DDR for other platforms.
> I'm sorry I'm confused. On what platforms do we want to have DDR by default?

As I understand it, we build all platforms with DDR enabled. It is off by default otherwise (for downstream projects).

I think what we want is to change `cmake/caches/Travis.cmake`, replacing the line
```
set(OMR_DDR ON CACHE BOOL "")
```
with
```
if(CMAKE_CROSSCOMPILING)
	set(OMR_DDR OFF CACHE BOOL "Enable DDR")
else()
	set(OMR_DDR ON CACHE BOOL "Enable DDR")
endif
```
That should have the effect of disabling DDR for cross-compiles.

BTW, 'Travis' is a misleading name for that file. The builds currently run on Azure, not TravisCI.

janvrany commented:
@genie-omr build riscv

janvrany commented:
Partial success - all tests exept porttest and comptest pass. Also DDR issues is not yet addressed. 

One thing I realized is that we use hardcoded paths to sysroot. @AdamBrousseau: would it be possible to define an environment variable, say 
`CROSS_SYSROOT_RISCV64` on RISC-V cross-compile hosts so this can be configured at Jenkins level? Since now it forces everyone to install things into the same locations.
Path to qemu is also hardcoded.

fjeremic commented:
@janvrany awesome stuff. Could you copy f0500bbaa4d34aa5384e612fd6586a7d78747a04 from #5753? That will fix the `comptest` failures. I've tested it to work already.

janvrany commented:
@fjeremic Sure, will make another pass tomorrow morning.

janvrany commented:
@genie-omr build riscv

janvrany commented:
@genie-omr build riscv

janvrany commented:
@keithc-ca 
> I think what we want is to change cmake/caches/Travis.cmake, replacing the line

The issue is that (apparently) one cannot do that in a cache file (in file passed to CMake's `-C`). IIUC, this is loaded before anything else so at that point, CMake does not know it is cross-compiling. I just tried that and it does not work as expected. 

I have no idea how to implement required logic so we may well need to drop commit dfea0b1 altogether and and `-DOMR_DDR=OFF` to cross-compile builds. Has anyone an idea?

janvrany commented:
This is interesting failure I do not see on my system:

```
07:46:38  14: [----------] 21 tests from PortStrTest
07:46:38  14: /home/jenkins/workspace/Build/fvtest/porttest/omrstrTest.cpp line  691: omrstr_test4  %uid was not substituted
07:46:38  14: 
07:46:38  14: 		LastErrorNumber: 2
07:46:38  14: 		LastErrorMessage: Insufficient buffer memory while attempting to load a shared library
07:46:38  14: 
07:46:38  14: /home/jenkins/workspace/Build/fvtest/porttest/testHelpers.cpp:109: Failure
07:46:38  14: Value of: 0 == numberFailedTestsInComponent
07:46:38  14:   Actual: false
07:46:38  14: Expected: true
07:46:38  14: Test failed!
07:46:38  14: [  FAILED  ] PortStrTest.str_test4 (271 ms)
```

keithc-ca commented:
I suggest we only need to worry about automated builds for the time being.
Modifying the `cmake` command-line in [Build-linux_riscv64_cross](https://github.com/eclipse/omr/blob/master/buildenv/jenkins/jobs/builds/Build-linux_riscv64_cross#L37-L43) should suffice.

janvrany commented:
@genie-omr build riscv

fjeremic commented on fvtest/porttest/CMakeLists.txt:
```diff
@@ -136,16 +136,23 @@ if(OMR_HOST_OS STREQUAL "aix")
 	)
 endif()
 
+# Some tests are known to file when running under simulator such as QEMU.
+# Disable them here.
+set(porttest_filter "*")
+if(CMAKE_CROSSCOMPILING)
+	set(porttest_filter "-PortFileTest2.*:PortMmapTest.*:PortSigTest.*:PortSysinfoTest.*")
+endif()
+
 if (NOT OMR_HOST_OS STREQUAL "aix" AND NOT OMR_HOST_OS STREQUAL "zos")
```
Honestly I don't think we should really care too much about cross compiled porttests. I would just add `AND NOT CMAKE_CROSSCOMPILING` to the `if `on line 146 and not run it at all. We get coverage for port test on other platforms anyway.

AdamBrousseau commented:
>define an environment variable, say
CROSS_SYSROOT_RISCV64 on RISC-V cross-compile hosts so this can be configured at Jenkins level?

Sorry I just saw this. Is this still required? I can add it to the node configuration on Jenkins. What value should I set it to?

janvrany commented:
@genie-omr build riscv

janvrany commented:
@genie-omr build riscv

fjeremic commented:
@genie-omr build riscv

janvrany commented:
@genie-omr build riscv

janvrany commented:
@genie-omr build riscv

janvrany commented:
OK, after a bunch of random failures, finally a green build.

janvrany commented:
@AdamBrousseau 

> Sorry I just saw this. Is this still required? I can add it to the node configuration on Jenkins. What value should I set it to?

Yes, that would be great. Can you set environment variable `CROSS_SYSROOT_RISCV64` to value `/home/jenkins/riscv-debian/rootfs`

It'd also be nice having `qemu-riscv64` in `PATH` so there's no need to hardcode path to QEMU executable. 
It should be OK to add `/home/jenkins/qemu/build` to `PATH` on the build host. 

Thanks a lot!

fjeremic commented:
Thanks for this work @janvrany. Once we get the envvars set, I'm happy to go forward with this change. I'll also enable the cross compile pipeline to run as part of "build all". It takes roughly ~50 minutes which is manageable. We can enable it and see how it goes. We may need additional cross compile machines but that's easy now that we know what to do.

One last request that I'll make is to squash some of the not so useful commits in the PR (the "oops" one in particular) as they may not add value in the history of the project.

AdamBrousseau commented:
Sorry, I set the vars yesterday and got distracted trying to test them in a build. My permissions to launch the PR builds seems to be flaky. I have set the PATH addition in the .profile and set the other var on the Node config. Please test and let me know if there's any issue.

janvrany commented:
@genie-omr build riscv

janvrany commented:
@AdamBrousseau: Thanks! 
The `PATH` however does not seem to contain the path to QEMU:
```
19:11:18  PATH=/usr/lib/ccache/:/usr/local/bin:/usr/bin:/bin:/usr/games
```
Adding it to `.profile` IIUC won't work in this case as shell spawned by Jenkins is not an interactive shell therefore `~/.profile` is not read (nor `~/.bashrc`).

janvrany commented:
@fjeremic Perfect. Sure, I'll tidy up the history, that was always the plan - once builds are green and everyone's happy.

AdamBrousseau commented:
>Adding it to .profile IIUC won't work

Right, this one always bites me. Can we add it to the omrbuild.groovy in the environment section? PATH+QEMU

janvrany commented:
@genie-omr build riscv

janvrany commented:
@AdamBrousseau, just looking at output of `printenv` in https://ci.eclipse.org/omr/job/PullRequest-linux_riscv64_cross/22/console, I do not see `CROSS_SYSROOT_RISCV64` there. It should be there no?

AdamBrousseau commented:
I think it was lost on the node during the final change over of the ci migration. I have added it back but it will get wiped again next time the instance is rebooted or redeployed. Please give it another try, if it is confirmed working then I'll open the PR to have it added permanently.

janvrany commented:
@genie-omr build riscv

janvrany commented:
@genie-omr build riscv

janvrany commented:
@genie-omr build riscv

janvrany commented:
@genie-omr build riscv

janvrany commented:
@genie-omr build riscv

janvrany commented:
@genie-omr build riscv

janvrany commented:
@genie-omr build riscv

janvrany commented:
@AdamBrousseau 
> I have added it back but it will get wiped again next time the instance is rebooted or redeployed

I think this is what has just happened - in https://ci.eclipse.org/omr/job/PullRequest-linux_riscv64_cross/28/ the variable seems to be gone again. May I kindly ask you to add it again? 

@fjeremic, @AdamBrousseau
I feel (hope) we're nearly done here. I have just pushed new version with commits squashed and reordered to "make more sense". If you have any comments, please let me know.

fjeremic commented:
> I have added it back but it will get wiped again next time the instance is rebooted or redeployed

Is there a way to avoid this? We get restarts fairly frequently, so it means the build will be busted every time?

AdamBrousseau commented:
Yes I just need to open a PR to have it added permanently. I assume when I had it added it was working so I'll do that now.

janvrany commented:
@genie-omr build riscv

AdamBrousseau commented:
PR is merged and node has the var again.

fjeremic reviewed (approved):
I'm happy with the changes here. Would like @keithc-ca to sign off as well give he has had previous reviews.

fjeremic commented:
@genie-omr build all

fjeremic commented:
Thanks so much @janvrany!
